### PR TITLE
Migrate rustc_mir_build's non_exhaustive_match diagnostic

### PR DESCRIPTION
**Pull request by mejrs**
This also changes the diagnostic somewhat, I hope that's OK

### CODE REVIEW

**Comment by compiler-errors** [comment]
I'm generally in favor for the primary message being tweaked along these lines -- but could you perhaps split that out into a separate PR that edits the message without doing the diagnostics migration (i.e. using the old diagnostics machinery), so others can chime in on the wording and presentation?

**Comment by mejrs** [comment]
I realized this PR would be hard to follow, which is why I did this change in multiple commits. Each of those makes sense on its own, which I hope makes the reviewing easier.

**Comment by compiler-errors** [comment]
Well it still needs fixing regardless -- was just trying to make it possible to get some of these changes (at least their presentation) landed before you were able to debug the rest of it.

@rustbot author

**Comment by est31** [comment]
FTR this is what rust-analyzer prints. Like the current error it's straight to the point, which I prefer:

Screenshot_20230222_171006

From what I can tell though the rust-analyzer error is native to them though, so this PR won't disrupt that IDE error.

I think if the set of patterns not handled becomes very complex then it makes sense to switch to the suggested error scheme and only explain the missing patterns in the report below, but otherwise I prefer the older way where the missing patterns are right at the top.

File: tests/ui/closures/2229_closure_analysis/match/issue-88331.stderr [hunk 1]
```diff
@@ -10,13 +10,14 @@ note: `Opcode` defined here
 LL | pub struct Opcode(pub u8);
    |            ^^^^^^
    = note: the matched value is of type `Opcode`
-help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms
+help: ensure that all possible cases are being handled by adding more match arms
    |
 LL ~         Opcode::OP1 => unimplemented!(),
-LL ~         Opcode(0_u8) | Opcode(2_u8..=u8::MAX) => todo!(),
+LL +         Opcode(0_u8) => { todo!() },
```
Comment (est31) [inline tests/ui/closures/2229_closure_analysis/match/issue-88331.stderr:16]: Note that while this is also my personal preference, to put a `,` after blocks in match arms, rustfmt's defaults are to omit it.

Comment (compiler-errors) [reply tests/ui/closures/2229_closure_analysis/match/issue-88331.stderr:16]: I agree this should be reverted. If you want to make this change separately, break it out into another PR.

Comment (mejrs) [reply tests/ui/closures/2229_closure_analysis/match/issue-88331.stderr]: I've reverted the commas.

**Comment by mejrs** [comment]
oops i think i forgot this:

@rustbot ready

**Comment by compiler-errors** [comment]
@rustbot author -- needs a rebase

feel free to ping me when this is ready for a review, sorry for just seeing this now

**Comment by JohnCSimon** [comment]
Ping from triage: 
@mejrs can you please address the merge conflicts, rebase etc?

**Comment by mejrs** [comment]
I will, my apologies for the delay :)

**Comment by JohnCSimon** [comment]
@mejrs 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
